crates/token/src/lib.rs

```diff
@@ -108,21 +101,114 @@ impl Transfer {
+        Self {
+            shielded_section_hash: Some(hash),
+            ..Self::default()
+        }
+    }
+
+    /// Set the key to the given amount
+    fn set<K: Ord>(
+        map: &mut BTreeMap<K, DenominatedAmount>,
+        key: K,
+        val: DenominatedAmount,
+    ) {
+        if val.is_zero() {
+            // Zero entries do not need to be present
+            map.remove(&key);
+        } else {
+            map.insert(key, val);
+        }
+    }
+
+    /// Credit the given account
+    pub fn credit(
+        mut self,
+        owner: Address,
+        token: Address,
+        amount: DenominatedAmount,
+    ) -> Option<Self> {
+        let account = Account { owner, token };
+        let zero = DenominatedAmount::new(Amount::zero(), amount.denom());
+        let source_amount = *self.sources.get(&account).unwrap_or(&zero);
+        let target_amount = *self.targets.get(&account).unwrap_or(&zero);
+        // If this account is already a target, then reduce the target
+        if amount < target_amount {
+            // Account still gets net increase
+            Self::set(
+                &mut self.targets,
+                account,
+                target_amount.checked_sub(amount)?,
+            );
+        } else {
+            // Account now actually gets a net decrease
+            self.targets.remove(&account);
+            let new_amt = source_amount
+                .checked_add(amount.checked_sub(target_amount)?)?;
+            Self::set(&mut self.sources, account, new_amt);
+        }
+        Some(self)
+    }
+
+    /// Debit the given account
+    pub fn debit(
+        mut self,
+        owner: Address,
+        token: Address,
+        amount: DenominatedAmount,
+    ) -> Option<Self> {
+        let account = Account { owner, token };
+        let zero = DenominatedAmount::new(Amount::zero(), amount.denom());
+        let source_amount = *self.sources.get(&account).unwrap_or(&zero);
+        let target_amount = *self.targets.get(&account).unwrap_or(&zero);
+        // If this account is already a source, then reduce the source
+        if amount < source_amount {
+            // Account still gets net decrease
+            Self::set(
+                &mut self.sources,
+                account,
+                source_amount.checked_sub(amount)?,
+            );
+        } else {
+            // Account now actually gets a net increase
+            self.sources.remove(&account);
+            let new_amt = target_amount
+                .checked_add(amount.checked_sub(source_amount)?)?;
+            Self::set(&mut self.targets, account, new_amt);
+        }
+        Some(self)
+    }
+
+    /// Transfer assets between accounts
+    pub fn transfer(
+        self,
+        source: Address,
+        target: Address,
+        token: Address,
+        amount: DenominatedAmount,
+    ) -> Option<Self> {
+        self.credit(source, token.clone(), amount)?
+            .debit(target, token, amount)
```

Comment:
isn't this other way around? I.e. transfer debits the source and credits the target